src/main/java/jenkins/scm/api/SCMSource.java

```diff
@@ -131,14 +133,65 @@ public boolean equals(Object obj) {
      *
      * @param id the id or {@code null}.
      */
+    @Deprecated
     protected SCMSource(@CheckForNull String id) {
         this.id = id == null ? UUID.randomUUID().toString() : id;
     }
 
     /**
-     * The ID of this source. The ID is not related to anything at all.
+     * Constructor.
+     */
+    protected SCMSource() {
+    }
+
+    /**
+     * Sets the ID that is used to ensure this {@link SCMSource} can be retrieved from its {@link SCMSourceOwner}
+     * provided that this {@link SCMSource} does not already {@link #hasId()}.
+     * <p>
+     * Note this is a <strong>one-shot</strong> setter. If {@link #getId()} is called first, then its value will stick,
+     * otherwise the first call to {@link #setId(String)} will stick.
+     *
+     * @param id the ID.
```

Comment:
This would be a good place to document that it is an opaque token, expected to be unique within one owner, etc.